internal/vcs/metadata.go

```diff
@@ -223,12 +288,332 @@ func (f *metadataFetcher) getLocalGitMetadata(path string) (Metadata, error) {
+		// However, calling the API requires authentication with variables that we don't
+		// have access to in the pipeline (e.g. username, password for basic auth).
+		Title:  "",
+		Author: "",
+	}
+
+	return m, nil
+}
+
+func (f *metadataFetcher) getCircleCIMetadata(path string) (Metadata, error) {
+	m, err := f.getLocalGitMetadata(path)
+	if err != nil {
+		return m, fmt.Errorf("circle CI metadata error, could not fetch initial metadata from local git %w", err)
+	}
+
+	m.Pipeline = &Pipeline{ID: os.Getenv("CIRCLE_WORKFLOW_ID")}
+	m.PullRequest = &PullRequest{
+		VCSProvider:  "circleci",
+		ID:           os.Getenv("CIRCLE_PR_NUMBER"),
+		SourceBranch: os.Getenv("CIRCLE_BRANCH"),
+		URL:          os.Getenv("CIRCLE_PULL_REQUEST"),
+
+		// we're unable to fetch these without calling the GitHub, Bitbucket or Gitlab API respectively.
+		// Calling the API requires authentication with variables that we don't have access to in the pipeline.
+		// e.g. a token or username, password for basic auth.
+		Title:      "",
+		Author:     "",
+		BaseBranch: "",
+	}
+
+	return m, nil
+}
+
+func (f *metadataFetcher) getAtlantisMetadata(path string) (Metadata, error) {
+	m, err := f.getLocalGitMetadata(path)
+	if err != nil {
+		return m, fmt.Errorf("atlantis metadata error, could not fetch initial metadata from local git %w", err)
+	}
+
+	// Atlantis doesn't expose any unique identifiers that we can use to distinguish a pipeline run.
+	m.Pipeline = &Pipeline{ID: ""}
+
+	m.PullRequest = &PullRequest{
+		VCSProvider:  "atlantis",
+		ID:           os.Getenv("PULL_NUM"),
+		Author:       os.Getenv("PULL_AUTHOR"),
+		SourceBranch: os.Getenv("HEAD_BRANCH_NAME"),
+		BaseBranch:   os.Getenv("BASE_BRANCH_NAME"),
+		// Atlantis doesn't provide any indication of which VCS system triggers a build.
+		// So we build the URL using the remote that the local git config points to.
+		URL: getAtlantisPullRequestURL(m.Remote),
+
+		// We're unable to fetch title without calling the GitHub, Bitbucket or Gitlab API respectively.
+		// This could be possible using the Atlantis Repo tokens, but it has been left out for now.
+		Title: "",
+	}
+
+	return m, nil
+}
+
+func getAtlantisPullRequestURL(remote Remote) string {
+	owner := os.Getenv("BASE_REPO_OWNER")
+	project := os.Getenv("BASE_REPO_NAME")
+	pullNumber := os.Getenv("PULL_NUM")
+
+	if strings.Contains(remote.Host, "github") {
+		return fmt.Sprintf("https://github.com/%s/%s/pull/%s", owner, project, pullNumber)
+	}
+
+	if strings.Contains(remote.Host, "gitlab") {
+		return fmt.Sprintf("https://gitlab.com/%s/%s/-/merge_requests/%s", owner, project, pullNumber)
+	}
+
+	if strings.Contains(remote.Host, "azure") {
+		return fmt.Sprintf("https://dev.azure.com/%s/base/_git/%s/pullrequest/%s", owner, project, pullNumber)
+	}
+
+	if strings.Contains(remote.Host, "bitbucket") {
+		return fmt.Sprintf("https://bitbucket.org/%s/%s/pull-requests/%s", owner, project, pullNumber)
+	}
```

Comment:
This method would cover cloud/hosted versions of GitHub/Gitlab/Bitbucket, but it seems for Enterprise versions we need to check their respective env vars like `ATLANTIS_GH_HOSTNAME` or `ATLANTIS_BITBUCKET_BASE_URL`. Or maybe you tried that and it's not possible for some reason? I took the env var names from here https://www.runatlantis.io/docs/server-configuration.html#environment-variables.

Reply:
yup, nice catch, will update the URL string to use those vars if they exist.

Reply:
Ah crap, yeah I remember, these aren't actually available to us in the run environment. These are purely server config vars, they don't seem to be exposed to the `run` step. I guess we could try and infer the enterprise versions from the `remote.Host` value